Add head status and snapshot utxo to the Greetings message

v0d1ch opened the pull request:
fix #823 

## Why

User request to alter the `Greetings` message to add some extra information

## What
We decided to use _projections_ from `CQRS` terminology. This alows us to keep the internal model of some state and at the same time update this model on certain server outputs we are interested. Projections always need to use a function in shape of `model -> event -> model` since the event is what drives the internal model update. 

---

<!-- Tick off or strike-through / remove if not applicable -->
* [x] CHANGELOG updated
* [ ] Documentation updated
* [x] Added and/or updated haddocks
* [x] No new TODOs introduced or explained herafter

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -267,6 +272,51 @@ spec = describe "ServerSpec" $
+                    case KeyMap.lookup "headStatus" km of
+                      Just hs -> guard $ hs == toJSON Open
+                      _other -> Nothing
+                  _other -> Nothing
```
Make assertions a bit smaller so it's less noisy.

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -267,6 +272,51 @@ spec = describe "ServerSpec" $
+
+            outputs :: [ServerOutput SimpleTx] <- generate arbitrary -- `suchThat` (/= Greetings)
+            mapM_ sendOutput outputs
+            let lastHeadStatus = foldl' projectHeadStatus Idle outputs
```
I would not use the server-side code to test expectations on the client-side.

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -267,6 +272,51 @@ spec = describe "ServerSpec" $
+                case v of
+                  Aeson.Object km -> do
+                    case KeyMap.lookup "headStatus" km of
+                      Just hs -> guard $ hs == toJSON ReadyForFanout
```
Also: Do not use the server-side types.

ch1bo commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -237,3 +239,41 @@ prepareServerOutput ServerOutputConfig{txOutputFormat, utxoInSnapshot} response
+  arbitrary = genericArbitrary
+
+-- | Projection type used to alter/project the API output to suit client needs
+newtype Projection a = Projection {runProjection :: TVar IO a}
```
Could: go into a dedicated module.

ch1bo replied:
Should: not expose the inner TVar

ch1bo committed:
Draft the api.yaml definition for updated Greetings message

The API reference is what our users will care about, so we start here.

ch1bo committed:
Write a test for assering headStatus in Greetings

This is deliberatly using the JSON objects directly to not need and rely
on any server-side FromJSON instances in this test suite.

v0d1ch committed:
Introduce HeadStatus type

- Start to work on ServerSpec tests related to Greetings because we
wouldn't like to match on equality.

ch1bo committed:
Rewrite ServerSpec to use predicates for greetings

We do not want to match on equality as the headStatus is often unrelated
to the expectation of the tests.

v0d1ch committed:
Create a Projection abstraction and use it in the Server

The idea of such a projection is to have query-only replicas of the
original event stream (ServerOutput in our case) and provide access to
them (later) through an API.

v0d1ch committed:
Added projection function and fixed a HeadStatus constructor name

v0d1ch committed:
Implement the update of the projection

- Use real persistence in the ServerSpec to exercise loading from disk

v0d1ch committed:
Add haddocks and rename newProjection to mkProjection

v0d1ch committed:
Fix golden tests

v0d1ch committed:
Introduce a failing test for snapshotUtxo field in Greetings

v0d1ch committed:
Make the tests green again

v0d1ch committed:
Improve assertions further to have a failing test

- Send a SnapshotConfirmed message and expect the utxos to match in
snapshotUtxo field of Greetings

v0d1ch committed:
Make the tests pass again

v0d1ch committed:
Add a test to make sure loading from persistence works

v0d1ch committed:
Fix golden files again

v0d1ch committed:
Update the changelog and api.yaml

v0d1ch commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -267,9 +277,109 @@ spec = describe "ServerSpec" $
+              status <- waitMatch 5 conn $ \v -> v ^? key "headStatus"
+              status `shouldBe` Aeson.String "Idle"
+
+            -- TODO: do test commit outputs as well? (or just unit/property test the projection for details)
```
There is this todo I am unsure what to do about. Seems like our `projectHeadStatus` and `projectSnapshotUtxo` are not good fit for a property test? Or maybe I just suck at this and need a helping hand.

ch1bo replied:
I think we can just drop this. It is a "test that `HeadIsOpen` corresponds to `Open`" in disguise - which is not a helpful test.

We can even think of dropping 1-2 of the other the `sendOutput` + expectations blocks in this test. In the end, this is an integration test that kind of makes sure that the server output is aggregated into the `headStatus` field.

v0d1ch committed:
Include HeadIsOpen message in our snapshotUtxo projection

v0d1ch committed:
Update Projection module with updated import

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -234,6 +234,8 @@ components:
           - tag
           - me
+          - headStatus
+          - snapshotUtxo
```
Must: Update the api spec to indicate this is optional.

Drop this from the `required` list, as it may not be included (unless something else comes out here https://github.com/input-output-hk/hydra/issues/823#issuecomment-1521750111)

ch1bo commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -81,7 +81,7 @@ data ServerOutput tx
     -- node. Currently used for knowing what signing key the server uses (it
     -- only knows one).
-    Greetings {me :: Party}
+    Greetings {me :: Party, headStatus :: HeadStatus, snapshotUtxo :: UTxOType tx}
```
Should: model the utxo here as `Maybe (UTxOType tx)` and update the haddocs

Only the `Open` head status has a utxo set. Either move the utxo into the `Open` `HeadStatus` or make this an optional field. In the latter case we should check that the `ToJSON` instance does leave out the field if it is `Nothing`

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -267,9 +277,105 @@ spec = describe "ServerSpec" $
+                    { headId = HeadId "some-head-id"
+                    , snapshot = generatedSnapshot
+                    , signatures = mempty
+                    }
```
Why not `SnapshotConfirmed <$> arbitrary <*> arbitrary <*> arbitrary`?

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -267,9 +277,105 @@ spec = describe "ServerSpec" $
+                    , signatures = mempty
+                    }
+            let expectedUtxos =
+                  toJSON (Hydra.Snapshot.utxo $ Hydra.API.ServerOutput.snapshot snapShotConfirmedMessage)
```
Could: use pattern matches to avoid the qualified imports and leverage DuplicateRecordFields (we also prefer record field puns over using accessor functions)

```suggestion
            let SnapshotConfirmed {snapshot = Snapshot{utxo}} = snapshotConfirmedMessage 
                expectedUtxos = toJSON utxo
```

Alternatively, generate the utxo separately and just replace it in your `generatedSnapshot`.

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -267,9 +277,105 @@ spec = describe "ServerSpec" $
+            withAPIServer @SimpleTx "127.0.0.1" port alice apiPersistence tracer noop $ \_ -> do
+              withClient port "/?history=no" $ \conn -> do
+                status <- waitMatch 5 conn $ \v -> v ^? key "snapshotUtxo"
+                status `shouldBe` expectedUtxos
```
The headStatus and snapshotUtxo tests can be combined easily. They are both about the Greetings message, so that also sounds quite right to merge them.

Also there is quite some duplication between the restart and not restart tests, whereas the first part of the restart tests is exactly the same as the non-restart tests.

DRY please :)

ch1bo commented on hydra-node/golden/ReasonablySized (ServerOutput SimpleTx).json:
```diff
@@ -2236,9 +2271,11 @@
-                "vkey": "bca9150f8ede476548571de3e4dcad31a620f54085b307ebbbb95b3003684148"
+                "vkey": "05446c191ec3f312872da337aa59c92325b28de4ae150419838673f5e8396838"
             },
+            "snapshotUtxo": [],
```
Should: Not serialize empty array here when there are no utxo.

This is a nice example of why this looks unintuitive on the API.

ch1bo reviewed (changes requested):
Good job on this! I think we want to make the serialization leave out the field though, so let's change that.

Maybe even move the utxo into the `Open` `HeadStatus` which would make most sense on the API (irrespective how we implement it).

v0d1ch committed:
PR review changes

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -808,6 +812,16 @@ components:
+        - Initializing
+        - Open
+        - Closed
+        - ClosedAfterDeadline
```
@v0d1ch I'm sorry, but they actually wished `FanoutPossible` here and we might just use that? See #823 :)

v0d1ch replied:
No worries, I'll change it quickly.

v0d1ch replied:
...and done!

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -231,9 +231,11 @@ components:
         A friendly welcome message which tells a client something about the node. Currently used for knowing what Party the server embodies. This message produced whenever the hydra-node starts and clients should take consequence of seeing this. For example, we can assume no peers connected when we see 'Greetings'.
       payload:
         type: object
+        additionalProperties: true
```
Could: not allow additional properties (besides the ones listed in `properties`)

ch1bo commented on hydra-node/src/Hydra/API/Projection.hs:
```diff
@@ -0,0 +1,53 @@
+--    transform these outputs and add more (stateful) information (like the 'Hydra.API.ServerOutput.HeadStatus' model).
+--
+--    'Projection's always need to use a function in form of `(model -> event -> model)` where
+--    depending on event we are currently dealing with we might want to alter our existing model.
```
Should: move this to actual module documentation.

I have not looked at the generated html, but likely this is not getting picked up by haddock as module docs need to come just before the `module` keyword.

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -267,9 +273,72 @@ spec = describe "ServerSpec" $
+            mapM_ sendOutput [readyToFanoutMsg, snapShotConfirmedMsg']
+            waitForValue port $ \v -> do
+              guard $ status v == Just (Aeson.String "FanoutPossible")
+              guard $ snapshotUtxos v == Just expectedUtxos'
```
Could: not introduce variables that are only used once.

--- 

Could: not introduce variables for things we actually care about in the assertion

The expression `v ^? key "snapshotUtxo"` is very explicit about the fact that we want to use field `"snapshotUtxo"` and assert it's value. I do not think it's a problem, but even think it's nice to be explicit in tests sometimes (DRY the boilerplate, be explicit about your intentions)

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -267,9 +273,72 @@ spec = describe "ServerSpec" $
+            waitForValue port $ \v -> do
+              guard $ status v == Just (Aeson.String "Idle")
+              -- test that the 'snapshotUtxo' is excluded from json if there is no utxo
+              guard $ isNothing (v ^? key "snapshotUtxo")
```
Could: improve test assertions / errors by using `shouldBe` where possible.

I think it would be better to wait for any `Greetings` message and then do the assertions explicitly. Sorry I did not suggest this earlier.

```suggestion
            greeting1 <- waitForValue port $ \v ->
              if matchGreeting v
              then pure v
              else Nothing
            greeting1 ^. key "headStatus `shouldBe` Just (Aeson.String "Idle")
            -- test that the 'snapshotUtxo' is excluded from json if there is no utxo
            greeting1 ^. key "snapshotUtxo" `shouldBe` Nothing
```

There is likely a better way to do the match in `waitForValue` or to reduce this boilerplate.
However, this will be more explicit if one or the other field is not of the expected value.

ch1bo reviewed (approved):
Found some more (minor) things you _could_ do. The overall approach and execution of this is fine by me though!
